File: LeetCode/0106. Construct Binary Tree from Inorder and Postorder Traversal/Solution.java
/**
 * Created by Inno Fang on 2018/3/20.
 */

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

/**
 * 203 / 203 test cases passed.
 * Status: Accepted
 * Runtime: 43 ms
 */
public class Solution {

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        return buildTree(postorder, 0, postorder.length, inorder, 0, inorder.length);
    }

    private TreeNode buildTree(int[] postOrder, int postStart, int postEnd, int[] inOrder, int inStart, int inEnd) {
        if (postStart >= postEnd || inStart >= inEnd) return null;
        TreeNode node = new TreeNode(postOrder[postEnd - 1]);
        int i = inStart;
        for(; i < inEnd; i++) if (inOrder[i] == node.val) break;
        int dis = i - inStart;
        node.left = buildTree(postOrder, postStart, postStart + dis, inOrder, inStart, inStart + dis);
        node.right = buildTree(postOrder, postStart + dis, postEnd - 1, inOrder, inStart + dis + 1, inEnd);
        return node;
    }
}

File: LeetCode/0106. Construct Binary Tree from Inorder and Postorder Traversal/Solution.kt
/**
 * Created by Inno Fang on 2018/3/20.
 */

class TreeNode(var `val`: Int = 0) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}

/**
 * 203 / 203 test cases passed.
 * Status: Accepted
 * Runtime: 596 ms
 */
class Solution {
    fun buildTree(inorder: IntArray, postorder: IntArray): TreeNode? {
        if (inorder.isEmpty() || postorder.isEmpty()) return null
        val node = TreeNode(postorder.last())
        val idx = inorder.lastIndexOf(node.`val`)
        node.left = buildTree(inorder.sliceArray(0 until idx), postorder.sliceArray(0 until idx))
        node.right = buildTree(inorder.sliceArray(idx + 1..inorder.lastIndex), postorder.sliceArray(idx until postorder.lastIndex))
        return node
    }
}